runtime/lua/vim/func/_memoize.lua

```diff
@@ -33,31 +35,51 @@ local function resolve_hash(hash)
   return hash
 end
 
+--- @param weak boolean?
+--- @return table
+local create_cache = function(weak)
+  return setmetatable({}, {
+    __mode = weak ~= false and 'kv',
+  })
+end
+
 --- @generic F: function
 --- @param hash integer|string|fun(...): any
 --- @param fn F
---- @param strong? boolean
+--- @param weak? boolean
 --- @return F
-return function(hash, fn, strong)
+return function(hash, fn, weak)
   vim.validate({
     hash = { hash, { 'number', 'string', 'function' } },
     fn = { fn, 'function' },
+    weak = { weak, 'boolean', true },
   })
 
-  ---@type table<any,table<any,any>>
-  local cache = {}
-  if not strong then
-    setmetatable(cache, { __mode = 'kv' })
-  end
-
-  hash = resolve_hash(hash)
+  --- @type MemoObj
+  local obj = {
```

Comment:
Do callers need a way to ask `_memoize()` if something is already cached?

Reply:
The cache and hash function are exposed below.

Reply:
they're currently marked private, while `invalidate` / `clear` is not.

Reply:
The whole module is private.

Reply:
that doesn't explain why we are marking some of its functions as private and some not. presumably, as the internal interface matures, we will graduate the module. we wouldn't wait until that time to then decide which of these functions are private/public

Reply:
I think these comments are getting out of scope. The motivation here is to assist with other active PR's. Let's not over-engineer this, the current requirements are simple.

Reply:
> Do callers need a way to ask `_memoize()` if something (a `hash` key) is already cached?

I asked this question for my own understanding. There's nothing out of scope about that.

Reply:
> `e()` if something (a `hash` key) is already cached?

That should be pretty easy to do. Shall I add a TODO for it? I will say I think something like that would hopefully be rarely used/needed, since ideally callers would be able to set up the memoized function and then basically forget that it is a cache and just treat it like a normal function. 

Also this can also be done now by inspecting the hidden cache, e.g. with query.get: `:=vim.treesitter.query.get._cache` (and then if your entry exists in here you know it was already cached. not saying it should be done that way but at least it is hackable as of now even :laughing: )

Reply:
Based on [this](https://github.com/neovim/neovim/pull/30227#discussion_r1746381988) I guess it sounds like callers don't need a way to ask if a hash key is already cached, because there is no global memoize store. 

Not sure how I feel about this. It means that we don't have visbility into all these caches, and won't be able to give intuitive performance checks/warnings.

Reply:
I think we first should have any indication that these are possible performance issues; so far there has been none (but then this is not public API, although a number of plugins have their own local variant).

Reply:
>  I think we first should have any indication that these are possible performance issues;

Not in this case. Global long lived-caches will cause issues. They'll just be hidden. And "giving indication" will later be unnecessarily cumbersome if we don't have visibility, which is what I'm asking for (not in this PR, but eventually).